src/value.rs

```diff
@@ -0,0 +1,763 @@
+use pyo3::call::PyCallArgs;
+use pyo3::prelude::{PyAnyMethods, PyDictMethods, PyListMethods, PySetMethods};
+use pyo3::types::{PyBool, PyBytes, PyDict, PyFloat, PyInt, PyList, PySet, PyString, PyTuple};
+use pyo3::{Bound, IntoPyObject, Py, PyAny, PyResult, Python};
+use scylla_cql::_macro_internal::ColumnType::Native;
+use scylla_cql::_macro_internal::{
+    ColumnType, DeserializationError, DeserializeValue, FrameSlice, UdtIterator,
+};
+use scylla_cql::deserialize::value::{
+    BuiltinDeserializationErrorKind, FixedLengthBytesSequenceIterator, MapDeserializationErrorKind,
+    SetOrListDeserializationErrorKind, VectorDeserializationErrorKind, mk_deser_err,
+};
+use scylla_cql::frame::frame_errors::LowLevelDeserializationError;
+use scylla_cql::frame::response::result::{CollectionType, NativeType};
+use scylla_cql::frame::types;
+use scylla_cql::value::{
+    Counter, CqlDate, CqlDecimal, CqlDuration, CqlTime, CqlTimestamp, CqlTimeuuid, CqlVarint,
+};
+use std::convert::Infallible;
+use std::net::IpAddr;
+use std::sync::Arc;
+
+// NOTE: I intentionally do NOT use Scylla's `DeserializeValue` trait here.
+// The trait does not provide a `Python` argument, meaning that Python objects which
+// would have to be constructed inside `deserialize()` or deeper in recursion
+// would require acquiring the GIL separately for every column/element/UDT/List
+// during nested decoding.
+//
+// Using our own `PyDeserializeValue` trait keeps all Python conversions inside
+// a single `Python::with_gil(...)` boundary, allowing complex values (lists,
+// sets, UDTs, nested collections) to be deserialized without repeated GIL
+// acquisition. This avoids potential slowdown of acquiring GIL multiples times.
+//
+// If GIL-per-python object created would not be considered a problem and
+// eliminating the need of rewriting DeserializeValue trait would be more beneficial
+// We could switch back to using `DeserializeValue` trait.
+pub trait PyDeserializeValue<'frame, 'metadata, 'py>: Sized {
+    fn deserialize_py(
+        typ: &'metadata ColumnType<'metadata>,
+        v: Option<FrameSlice<'frame>>,
+        py: Python<'py>,
+    ) -> Result<PyDeserializedValue, DeserializationError>;
+}
+
+pub struct List<T> {
+    phantom_data: std::marker::PhantomData<T>,
+}
```

Comment:
Let's keep `pub` used only for items that are exposed to Python, and `pub(crate)` for items needed outside the module defining them. In other words, privacy of items should be as narrow as possible.
This is a general note, not only for those 2 items - I see this issue in other places, so please review your code and adjust.